lib/broadway_cloud_pub_sub/client_acknowledger.ex

```diff
@@ -0,0 +1,284 @@
+
+    * `on_failure` - The action to perform for failed messages. Default is `:ignore`.
+  """
+  @spec init(opts :: any) :: {:ok, t} | {:error, message :: binary}
+  def init(opts) do
+    with {:ok, client} <- validate(opts, :client),
+         {:ok, on_success} <- validate(opts, :on_success, :ack),
+         {:ok, on_failure} <- validate(opts, :on_failure, :ignore) do
+      {:ok,
+       %__MODULE__{
+         client: client,
+         on_failure: on_failure,
+         on_success: on_success
+       }}
+    end
+  end
+
+  @doc """
+  Creates a reference that can be passed to `acknowledger/2`.
+
+  The client configuration is stored in `Broadway.TermStorage`.
+  """
+  @spec ack_ref(config :: t, client_opts :: any) :: ack_ref
+  def ack_ref(config, client_opts) do
+    TermStorage.put(%{config | client_opts: client_opts})
+  end
+
+  @doc """
+  Returns an `acknowledger` to be put on a `Broadway.Message`.
+
+  ## Example
+
+      iex> BroadwayCloudPubSub.ClientAcknowledger.acknowledger("ackId", :ack_ref)
+      {BroadwayCloudPubSub.ClientAcknowledger, :ack_ref, %{ack_id: "ackId"}}
+
+  """
+  @spec acknowledger(ack_id :: Client.ack_id(), ack_ref) :: {__MODULE__, ack_ref, ack_data}
+  def acknowledger(ack_id, ack_ref) do
+    {__MODULE__, ack_ref, %{ack_id: ack_id}}
+  end
+
+  @impl Acknowledger
+  def ack(ack_ref, successful, failed) do
+    config = Broadway.TermStorage.get!(ack_ref)
+
+    success_actions = group_actions_ack_ids(successful, :on_success, config)
+    failure_actions = group_actions_ack_ids(failed, :on_failure, config)
+
+    success_actions
+    |> Map.merge(failure_actions, fn _, a, b -> a ++ b end)
+    |> ack_messages(config)
+
+    :ok
+  end
+
+  @impl Acknowledger
+  def configure(_ack_ref, ack_data, options) do
+    options = assert_valid_config!(options)
+    ack_data = Map.merge(ack_data, Map.new(options))
+    {:ok, ack_data}
+  end
+
+  defp assert_valid_config!(options) do
+    Enum.map(options, fn
+      {:on_success, value} -> {:on_success, validate_option!(:on_success, value)}
+      {:on_failure, value} -> {:on_failure, validate_option!(:on_failure, value)}
+      {other, _value} -> raise ArgumentError, "unsupported configure option #{inspect(other)}"
+    end)
+  end
+
+  defp group_actions_ack_ids(messages, key, config) do
+    Enum.group_by(messages, &group_acknowledger(&1, key, config), &extract_ack_id/1)
+  end
+
+  defp group_acknowledger(%{acknowledger: {_, _, ack_data}}, key, config) do
+    Map.get_lazy(ack_data, key, fn -> config_action(key, config) end)
+  end
+
+  defp config_action(:on_success, %{on_success: action}), do: action
+  defp config_action(:on_failure, %{on_failure: action}), do: action
+
+  defp extract_ack_id(message) do
+    {_, _, %{ack_id: ack_id}} = message.acknowledger
+    ack_id
+  end
+
+  defp ack_messages(actions_and_ids, config) do
+    Enum.each(actions_and_ids, fn {action, ack_ids} ->
+      ack_ids
+      |> Enum.chunk_every(3_000)
```

Comment:
This is based on a similar update in the nodejs lib -- perhaps it warrants adding some additional comments here 🤔 

googleapis/nodejs-pubsub#65

Reply:
:+1: for comments.

Reply:
Done.